segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,204 @@
+     *
+     * @return The current degree of parallelism.
+     */
+    int getCurrentParallelism() {
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
+     * Notes:
+     * * Only changes something if throughput changed significantly since the last time.
+     * * Changes the degree of parallelism in the same direction as throughput.
+     * ** Exception is if both throughput and the fill ratio are declining, in which case nothing is done.
+     *
+     * @return The degree of parallelism, whether it was updated or not.
+     */
+    int getOrUpdateParallelism() {
+        // Calculate the most recent throughput and Fill Ratio (of those items that have just been written).
+        final long time = this.timeSupplier.get();
+        final long elapsedMillis = (time - this.lastSnapshot.get().timeStamp) / AbstractTimer.NANOS_TO_MILLIS;
+        if (elapsedMillis >= STALE_MILLIS) {
+            // Too long since the last update; the data we have is no longer relevant. Reset.
+            resetSnapshot(time);
+        } else if (elapsedMillis >= UPDATE_FREQUENCY_MILLIS) {
+            // Enough time elapsed since the last update; update the snapshot.
+            updateSnapshot(time, elapsedMillis);
+        }
+
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the current snapshot based on the current data.
+     *
+     * @param time          The current time, in millis.
+     * @param elapsedMillis The elapsed time, in millis, since the last time this method ran.
+     */
+    private void updateSnapshot(long time, long elapsedMillis) {
+        final int recentCount;
+        final long recentLength;
+        synchronized (this) {
+            recentCount = this.recentWriteCount;
+            recentLength = this.recentTotalWrittenLength;
+            this.recentWriteCount = 0;
+            this.recentTotalWrittenLength = 0;
+        }
+
+        final double recentThroughput = (double) recentLength / elapsedMillis;
+        final double recentFillRatio = WriteQueue.calculateFillRatio(recentLength, recentCount);
+        final Snapshot lastSnapshot = this.lastSnapshot.get();
+        final double throughputDifference = recentThroughput - lastSnapshot.throughput;
+
+        int parallelism = lastSnapshot.parallelism;
+        if (Math.abs(throughputDifference / lastSnapshot.throughput) >= SIGNIFICANT_DIFFERENCE) {
+            // Throughput changed significantly.
+            // But first, check for a special case. If both Throughput and Fill Ratio are decreasing, then most likely the
+            // former is a result of the latter. As such, don't do anything, in that case.
+            boolean fillRatioDecreased = (recentFillRatio - lastSnapshot.fillRatio) <= -SIGNIFICANT_DIFFERENCE;
```

Comment:
This may not have the intended effect. If the parallelism is between 5 and 10 assuming that the incoming rate is constant, increasing the parallelism from 5 to 6 from 6 to 7 or from 7 to 8 will never trigger a drop in the fill rate of more than 25% because none of these changed 25% of the traffic. It might instead drop by 20% followed by another 16.6% followed by another 14.5% resulting in a very larger cumulative drop, but none of the individual drops was large enough to trigger the SIGNIFICANT_DIFFERENCE threshold.

Reply:
The fill rate is determined by a number of factors, such as client production rate, wire protocol (since it only has one append in flight per segment) and a heuristic in the `OperationProcessor` that decides how much to wait (and hence increase the fill rate) if BookKeeper log becomes too backed up (it compares the queue size with the current parallelism). So the math isn't as simple as you suggested. It's a more complex feedback mechanism that I tuned empirically based on several types of traffic patterns (including varying the traffic patterns to see how it adjusts). Maybe there is a way to express this as a pure mathematical formula, but so far this heuristic has proven to be the best performing of what I tried so far.